### ISSUE TITLE
Normalizing Flow When transform style is MAF(Masked Autoregressive Flow) 최종

### ISSUE THREAD

**Issue by GongsonChan** [GongsonChan/code#19]
## library
```
import pandas as pd
import torch
from sklearn.preprocessing import StandardScaler
from nflows.flows.base import Flow
from nflows.transforms.normalization import BatchNorm
from nflows.transforms.autoregressive import MaskedAffineAutoregressiveTransform
from nflows.transforms.permutations import ReversePermutation
from nflows.transforms.base import CompositeTransform
from nflows.distributions.normal import StandardNormal
import random
import numpy as np
```

## set.random seed
```
# Python random seed
random.seed(42)

# NumPy random seed
np.random.seed(42)
```

## 1. 데이터 불러오기
```
df = pd.read_csv('C:/Users/hhccw/OneDrive/바탕 화면/통공모/tgm2025_data.csv')  # 또는 df = your_dataframe
data = df[['lon', 'lat', 'time', 'intensity']].values
```

## 2. 스케일링
```
scaler = StandardScaler()
data_scaled = scaler.fit_transform(data)
```

## 3. Torch Tensor로 변환
`x = torch.tensor(data_scaled, dtype=torch.float32)`

## 4. 모델 구성 (4차원 데이터)
```
num_features = x.shape[1]
hidden_features = 7
num_layers = 3
```

## 5. Flow 모델 구성 - batchnorm 사용
```
transforms = []
for _ in range(num_layers):
    transforms.append(MaskedAffineAutoregressiveTransform(
        features=num_features, hidden_features=hidden_features))
    transforms.append(BatchNorm(num_features))
    transforms.append(ReversePermutation(features=num_features))

transform = CompositeTransform(transforms)
base_dist = StandardNormal(shape=[num_features])
flow = Flow(transform, base_dist)
```
## 파라미터 수 확인
```
total_params = sum(p.numel() for p in flow.parameters())
print(f"모형의 전체 파라미터 수: {total_params}")
```
> 모형의 전체 파라미터 수: 993

## 6. 모델 학습 - 반복횟수 2000
```
optimizer = torch.optim.Adam(flow.parameters(), lr=1e-3)
num_epochs = 2000

for epoch in range(num_epochs):
    optimizer.zero_grad()
    loss = -flow.log_prob(x).mean()  # Negative Log-Likelihood
    loss.backward()
    optimizer.step()
    if epoch % 100 == 0:
        print(f"Epoch {epoch}, Loss: {loss.item():.4f}")
```

## 7. 학습된 모델에서 log-MLE 계산
```
log_likelihood = flow.log_prob(x).sum().item()

print(f"총 Log-Likelihood: {log_likelihood:.4f}")
```
> 총 Log-Likelihood: -81926.9609
